checker/jdk/nullness/src/java/util/logging/LogRecord.java

```diff
@@ -0,0 +1,593 @@
+    public int getThreadID() {
+        return threadID;
+    }
+
+    /**
+     * Set an identifier for the thread where the message originated.
+     * @param threadID  the thread ID
+     */
+    public void setThreadID(int threadID) {
+        this.threadID = threadID;
+    }
+
+    /**
+     * Get event time in milliseconds since 1970.
+     *
+     * @return event time in millis since 1970
+     */
+    public long getMillis() {
+        return millis;
+    }
+
+    /**
+     * Set event time.
+     *
+     * @param millis event time in millis since 1970
+     */
+    public void setMillis(long millis) {
+        this.millis = millis;
+    }
+
+    /**
+     * Get any throwable associated with the log record.
+     * <p>
+     * If the event involved an exception, this will be the
+     * exception object. Otherwise null.
+     *
+     * @return a throwable
+     */
+    public @Nullable Throwable getThrown() {
+        return thrown;
+    }
+
+    /**
+     * Set a throwable associated with the log event.
+     *
+     * @param thrown  a throwable (may be null)
+     */
+    public void setThrown(@Nullable Throwable thrown) {
+        this.thrown = thrown;
+    }
+
+    private static final long serialVersionUID = 5372048053134512534L;
+
+    /**
+     * @serialData Default fields, followed by a two byte version number
+     * (major byte, followed by minor byte), followed by information on
+     * the log record parameter array.  If there is no parameter array,
+     * then -1 is written.  If there is a parameter array (possible of zero
+     * length) then the array length is written as an integer, followed
+     * by String values for each parameter.  If a parameter is null, then
+     * a null String is written.  Otherwise the output of Object.toString()
+     * is written.
+     */
+    @SuppressWarnings("dereference.of.nullable") // out.writeInt and out.writeObject do not affect parameters field. http://tinyurl.com/cfissue/984
+    private void writeObject(ObjectOutputStream out) throws IOException {
+        // We have to call defaultWriteObject first.
+        out.defaultWriteObject();
+
+        // Write our version number.
+        out.writeByte(1);
+        out.writeByte(0);
+        if (parameters == null) {
+            out.writeInt(-1);
+            return;
+        }
+        out.writeInt(parameters.length);
+        // Write string values for the parameters.
+        for (int i = 0; i < parameters.length; i++) {
+            if (parameters[i] == null) {
+                out.writeObject(null);
+            } else {
+                out.writeObject(parameters[i].toString());
+            }
+        }
+    }
+
+    @SuppressWarnings({"dereference.of.nullable", "argument.type.incompatible"}) // in.readObject does not affect parameters field. http://tinyurl.com/cfissue/984
+    // and ClassLoader.getSystemClassLoader() is @NonNull because  there will always be a SystemClassLoader present to load the classes of logging framework
```

Comment:
The comment about `ClassLoader.getSystemClassLoader()` should appear on [line 536](https://github.com/typetools/checker-framework/pull/2481/files#diff-3f0e5215c54358445cecd19a24665a8bR536), not here.
Please always write `@SuppressWarnings` in the [smallest scope possible](https://checkerframework.org/manual/#suppresswarnings-annotation-locations).
The `readObject` issue is more pervasive in the method, and its warning suppression can be on the entire method.

Reply:
I agree that there will always be a system class loader.  That does not imply that getSystemClassLoader will not return null in Java 8.

However, Java 9 has changed the contract of getSystemClassLoader so that it does not return null (it throws an exception instead; compare [JDK 8 Javadoc](https://docs.oracle.com/javase/8/docs/api/java/lang/ClassLoader.html#getSystemClassLoader--) with [JDK 9 Javadoc](https://docs.oracle.com/javase/9/docs/api/java/lang/ClassLoader.html#getSystemClassLoader--)).  

Therefore, I have [changed](https://github.com/typetools/checker-framework/commit/0b77004b1b2f999d40e64d7414378d3e1796585f) the annotation on its return type.  It isn't strictly true for the JDK that is being annotated, but we are switching to Java 9 soon, so I think it's OK.  This means you can remove the `@SuppressWarnings` related to `getSystemClassLoader()` here.

It's interesting that investigating a warning can lead to understanding of the JDK and its changes, and can lead to changed JDK annotations elsewhere.

Reply:
> I agree that there will always be a system class loader. That does not imply that getSystemClassLoader will not return null in Java 8.

I agree that it doesn't imply getSystemClassLoader will not return null in Java8, but since this class will be mostly used by libraries like slf4j and they will definitely have a ClassLoader(if they don't they will run right into NullPointerExceptiion).

>It's interesting that investigating a warning can lead to understanding of the JDK and its changes, and can lead to changed JDK annotations elsewhere.

Indeed, I didn't think that this small PR will lead to this interesting discovery. I  think Oracle just thought of the same thing and changed the contract in Java9.

@mernst Thanks for your help and all your valuable inputs.